src/v/pandaproxy/sharded_client_cache.cc

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/sharded_client_cache.h"
+
+#include "ssx/future-util.h"
+
+using namespace std::chrono_literals;
+
+namespace pandaproxy {
+
+static constexpr auto clean_timer_period = 10s;
+
+ss::future<> sharded_client_cache::start(
+  ss::smp_service_group sg,
+  YAML::Node const& cfg,
+  config::binding<size_t> max_size,
+  config::binding<std::chrono::milliseconds> keep_alive) {
+    _smp_opts = ss::smp_submit_to_options{sg};
+
+    _clean_timer.set_callback([this] {
+        ssx::spawn_with_gate(_gate, [this] {
+            return _cache.invoke_on_all(
+              _smp_opts,
+              [](kafka_client_cache& cache) { cache.clean_stale_clients(); });
+        });
+    });
+
+    _clean_timer.arm_periodic(clean_timer_period);
```

Comment:
> basically here instead of arm_periodic arm once. and then attach a finally block to invoke_on_all above which rearms the timer.

IIUC, you're suggesting to reset the callback with a nested `_clean_timer.set_callback([] {});`

I was under the impression that `timer.set_callback()` sets the callback one-time only.
So if I used your suggestion, the consequences would be:
1. Set a callback on the garbage collection time including a nested callback to reset
2. Timer is triggered and invokes garbage collection then the callback is reset
3. Timer is triggered again, at a later period, but callback was reset so no garbage collection

That is unfavorable. We want GC to run for the applications lifetime.

Reply:
I think @dotnwat means to call `.arm` instead of `.arm_periodic`. Once the work is done, it can call `.arm` again.

Reply:
@BenPope is correct. my comment wasn't clear enough, sorry about that.